### PR TITLE
Add committer field to changes

### PR DESCRIPTION
**Pull request by rajdeepbharati**
Closes https://github.com/buildbot/buildbot/issues/4709
## Contributor Checklist:

* [x] I have updated the unit tests
* [ ] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/changes/bitbucket.py [hunk 1]
```diff
@@ -116,6 +116,7 @@ def _processChanges(self, page):
                     # access additional information
                     author = pr['author']['display_name']
                     prlink = pr['links']['html']['href']
+                    committer = None
```
Comment (p12tic) [inline master/buildbot/changes/bitbucket.py]: We probably don't need this variable here, do we? Is it used later?

File: master/buildbot/changes/gitpoller.py [hunk 2]
```diff
@@ -285,6 +285,17 @@ def process(git_output):
             return git_output
         return d
 
+    def _get_commit_committer(self, rev):
```
Comment (p12tic) [inline master/buildbot/changes/gitpoller.py]: Could we use @defer.inlineCallbacks for new code?

File: master/buildbot/changes/svnpoller.py [hunk 3]
```diff
@@ -335,6 +335,7 @@ def create_changes(self, new_logentries):
 
             log.msg("Adding change revision %s" % (revision,))
             author = self._get_text(el, "author")
+            committer = None
```
Comment (p12tic) [inline master/buildbot/changes/svnpoller.py]: Is this variable used later?

File: master/buildbot/db/changes.py [hunk 4]
```diff
@@ -56,7 +56,7 @@ def thd(conn):
     def addChange(self, author=None, files=None, comments=None, is_dir=None,
                   revision=None, when_timestamp=None, branch=None,
                   category=None, revlink='', properties=None, repository='', codebase='',
-                  project='', uid=None):
+                  project='', uid=None, committer=None):
```
Comment (p12tic) [inline master/buildbot/db/changes.py]: I think for consistency let's put this the committer parameter just after author in all commits.

File: master/buildbot/test/fake/fakedb.py [hunk 5]
```diff
@@ -973,6 +975,7 @@ def fakeAddChangeInstance(self, change):
         row = dict(
             changeid=changeid,
             author=change.who,
+            # committer=change.committer,
```
Comment (p12tic) [inline master/buildbot/test/fake/fakedb.py]: Something is missing here?

File: master/buildbot/db/migrate/versions/053_add_changes_committer.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import sqlalchemy as sa
+
+from buildbot.util import sautils
+
+
+def upgrade(migrate_engine):
+    metadata = sa.MetaData()
+    metadata.bind = migrate_engine
+    changes_table = sautils.Table('changes', metadata, autoload=True)
+    committer = sa.Column('committer', sa.String(255), nullable=True, server_default='tommy')
```
Comment (tardyp) [inline master/buildbot/db/migrate/versions/053_add_changes_committer.py]: really? tommy? Nah.. shall be None

File: master/buildbot/test/unit/test_db_migrate_versions_053_add_changes_committer.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            metadata.bind = conn
+
+            changes = sautils.Table('changes', metadata, autoload=True)
+            self.assertIsInstance(changes.c.committer.type, sa.String)
+
+            q = sa.select([changes.c.author, changes.c.committer])
+            num_rows = 0
+            for row in conn.execute(q):
+                # verify that the default value was set correctly
+                self.assertEqual(row.committer, 'tommy')
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_migrate_versions_053_add_changes_committer.py]: shall be None or 'warner'

Comment (p12tic) [reply master/buildbot/test/unit/test_db_migrate_versions_053_add_changes_committer.py]: Better be None, I will write why below.

**Comment by p12tic** [comment]
I think one thing to keep in mind is that if we make the committer field is optional (or nullable), then it must be documented as such everywhere so that there's no confusion among the users of the DB or APIs. Additionally, the optionality means that the value None means it's unknown, thus we must not fill the value with any non-None values by default.

**Comment by p12tic** [comment]
I'm impressed too. There's only several minor issues with this PR. Congratulations!

File: master/buildbot/changes/gitpoller.py (hunk 2 above)
Comment (rajdeepbharati) [reply master/buildbot/changes/gitpoller.py]: Not sure how to do it (I'm not very familiar with twisted). Could you tell me?

File: master/buildbot/changes/bitbucket.py (hunk 1 above)
Comment (rajdeepbharati) [reply master/buildbot/changes/bitbucket.py]: Removed

File: master/buildbot/db/changes.py (hunk 4 above)
Comment (rajdeepbharati) [reply master/buildbot/db/changes.py]: Done

File: master/buildbot/test/unit/test_db_migrate_versions_053_add_changes_committer.py (hunk 7 above)
Comment (rajdeepbharati) [reply master/buildbot/test/unit/test_db_migrate_versions_053_add_changes_committer.py]: Done

File: master/buildbot/test/fake/fakedb.py (hunk 5 above)
Comment (rajdeepbharati) [reply master/buildbot/test/fake/fakedb.py]: I deleted that line (if I uncomment it, I get an error something like: `FakeChange object has not attribute committer`)

**Comment by rajdeepbharati** [comment]
https://buildbot.buildbot.net/#/builders/44/builds/601 this is failing because pylint isn't accepting `committer` as a valid english word
https://buildbot.buildbot.net/#/builders/23/builds/870 is passing locally (but I don't know why it didn't pass here)

**Comment by p12tic** [comment]
You can add new words to the whitelist at common/code_spelling_ignore_words.txt. As for the other failure, there's chance it's just an instability, I've launched that test again.

File: master/buildbot/changes/gitpoller.py (hunk 2 above)
Comment (tardyp) [reply master/buildbot/changes/gitpoller.py]: http://lmgtfy.com/?q=twisted+inlinecallbacks (2nd answer is the tutorial)

+ there are tons of examples in the buildbot code

Comment (rajdeepbharati) [reply master/buildbot/changes/gitpoller.py]: @tardyp @p12tic Is this the right way?
```py
@defer.inlineCallbacks
def _get_commit_committer(self, rev):
    args = ['--no-walk', r'--format=%aN <%aE>', rev, '--']
    try:
        res = yield self._dovccmd('log', args, path=self.workdir)
    except:
        raise EnvironmentError('could not get commit committer for rev')
    return res
```

**Comment by rajdeepbharati** [comment]
I have updated it

File: master/buildbot/changes/gitpoller.py [hunk 8]
```diff
@@ -285,6 +285,15 @@ def process(git_output):
             return git_output
         return d
 
+    @defer.inlineCallbacks
+    def _get_commit_committer(self, rev):
+        args = ['--no-walk', r'--format=%aN <%aE>', rev, '--']
+        res = yield self._dovccmd('log', args, path=self.workdir)
+        if res:
```
Comment (tardyp) [inline master/buildbot/changes/gitpoller.py]:
```suggestion
        if not res:
              raise EnvironmentError('could not get commit committer for rev')
        return res
```

File: master/buildbot/changes/gitpoller.py [hunk 9]
```diff
@@ -340,6 +349,7 @@ def _process_changes(self, newRev, branch):
             dl = defer.DeferredList([
                 self._get_commit_timestamp(rev),
                 self._get_commit_author(rev),
+                self._get_commit_committer(rev),
                 self._get_commit_files(rev),
                 self._get_commit_comments(rev),
```
Comment (tardyp) [inline master/buildbot/changes/gitpoller.py]: Probably we could get timestamp, author and commit in only one command but this is not a big deal

Comment (rajdeepbharati) [reply master/buildbot/changes/gitpoller.py]: Do you want me to combine them?

**Comment by rajdeepbharati** [comment]
codecov isn't accepting it

File: master/buildbot/changes/gitpoller.py (hunk 9 above)
Comment (p12tic) [reply master/buildbot/changes/gitpoller.py]: I think it's worth doing that as a separate PR. The less things in a single PR the less chance that it'll break.

**Comment by p12tic** [comment]
There's been a pull request updating tests in related area, I'll rebase the changes and give CI another chance to run.

**Comment by p12tic** [comment]
@rajdeepbharati Just realized master/docs/developer/database.rst needs updating.

**Comment by rajdeepbharati** [comment]
@p12tic Oops... what needs to be done?

**Comment by p12tic** [comment]
Update it in a next PR :) Just add the description of the new field and mention that it's optional.

File: master/docs/developer/database.rst [hunk 10]
```diff
@@ -775,6 +775,7 @@ changes
     * ``changeid`` (the ID of this change)
     * ``parent_changeids`` (list of ID; change's parents)
     * ``author`` (unicode; the author of the change)
+    * ``committer`` (unicode; the committer of the change)
```
Comment (rajdeepbharati) [inline master/docs/developer/database.rst]: @p12tic Already added, but I haven't mentioned that it's optional

Comment (p12tic) [reply master/docs/developer/database.rst]: Hm, sorry about that then. I'm wondering how I could miss it...

**Pull request merged**